src/SqlAst/ParserInference.php

```diff
@@ -61,27 +64,57 @@ public function narrowResultType(string $queryString, ConstantArrayType $resultT
                     $fromName = $from->getTable()->getName();
                     $fromTable = $this->schemaReflection->getTable($fromName);
                 } elseif ($from instanceof Join) {
-                    if ($fromTable === null) {
-                        $fromName = $from->getLeft()->getTable()->getName();
-                        $fromTable = $this->schemaReflection->getTable($fromName);
-                    }
-                    // @phpstan-ignore-next-line
-                    $joinName = $from->getRight()->getTable()->getName();
-                    $joinedTable = $this->schemaReflection->getTable($joinName);
+                    while (1) {
+                        if ($from->getCondition() === null) {
+                            if (QueryReflection::getRuntimeConfiguration()->isDebugEnabled()) {
+                                throw new UnresolvableQueryMixedTypeException('Cannot narrow down types null join conditions: ' . $queryString);
+                            }
+
+                            return $resultType;
+                        }
+
+                        if ($from->getRight() instanceof TableReferenceSubquery || $from->getLeft() instanceof TableReferenceSubquery) {
+                            if (QueryReflection::getRuntimeConfiguration()->isDebugEnabled()) {
+                                throw new UnresolvableQueryMixedTypeException('Cannot narrow down types for SQLs with subqueries: ' . $queryString);
+                            }
+
+                            return $resultType;
+                        }
+
+                        if ($from instanceof InnerJoin && $from->isCrossJoin()) {
+                            if (QueryReflection::getRuntimeConfiguration()->isDebugEnabled()) {
+                                throw new UnresolvableQueryMixedTypeException('Cannot narrow down types for cross joins: ' . $queryString);
+                            }
+
+                            return $resultType;
+                        }
```

Comment:
why do we need these early returns?

Reply:
it cannot collect proper data about joins and the columns hence it gives up on narrowing down the type

Reply:
but would it make a difference for the added test-cases if we just drop it here. wouldn't it give up after the loop then?

Reply:
without the early return, it wouldnt work, the tests would fail

```
1) Error
The data provider specified for staabm\PHPStanDba\Tests\DbaInferenceTest::testFileAsserts is invalid.
Error: Call to undefined method SqlFtw\Sql\Dml\TableReference\TableReferenceSubquery::getTable()
/Users/jakubvojacek/Sites/phpstan-dba/src/SqlAst/ParserInference.php:96
/Users/jakubvojacek/Sites/phpstan-dba/src/QueryReflection/QueryReflection.php:135
/Users/jakubvojacek/Sites/phpstan-dba/src/PdoReflection/PdoStatementReflection.php:88
/Users/jakubvojacek/Sites/phpstan-dba/src/Extensions/PdoQueryDynamicReturnTypeExtension.php:106
```

Reply:
but that will trigger phpstan internal error and the analysis will fail and phpstan cache wont be saved,right?I originally had the exception there but it was so annoying when running on my codebase that I decided to hide it behind debug flag.

Reply:
I think it should work when we create a new subclass of `UnresolvableQueryException`

similar to how we did with https://github.com/staabm/phpstan-dba/blob/main/src/UnresolvableQueryMixedTypeException.php which seems to work for you atm

Reply:
currently

Reply:
e.g. `UnresolvableAstInQueryException`

Reply:
so you want me to rename `UnresolvableQueryMixedTypeException` to `UnresolvableAstInQueryException` on those three places that it thrown in ParserInference right?

I checked where UnresolvableQueryMixedTypeException was used and its only on one place and its after the `QueryReflection::getRuntimeConfiguration()->isDebugEnabled()` condition as well

Reply:
nope, I want you to create a new exception class and use it here in this few places of this PR, but leave `UnresolvableQueryMixedTypeException`  as is across the codebase